File: Python/csv2html/downloader.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
import shutil
import urllib

from io import IO
from reader import Reader
from table import Table

class Downloader(IO):
	"""ダウンローダ：CSVファイル・画像ファイル・サムネイル画像ファイルをダウンロードする。"""

	def __init__(self, input_table):
		"""ダウンローダのコンストラクタ。"""

		super(Downloader, self).__init__(input_table)

		return

	def download_csv(self):
		"""情報を記したCSVファイルをダウンロードする。"""
		csv_url = self.attributes().csv_url()
		print("From: " + csv_url)
		filename = os.path.join(self.attributes().base_directory(), csv_url.split('/')[-1])
		print("To: " + filename)
		if os.path.exists(filename) and os.path.isfile(filename):
			pass
		else:
			urllib.urlretrieve(csv_url, filename)
		return

	def download_images(self, image_filenames):
		"""画像ファイル群または縮小画像ファイル群をダウンロードする。"""
		for image_file in image_filenames:
			image_url = self.attributes().base_url() + image_file
			print("From: " + image_url)
			image_path = os.path.join(self.attributes().base_directory(), image_file)
			print("To: " + image_path)
			urllib.urlretrieve(image_url,  image_path)
		return

	def perform(self):
		"""すべて（情報を記したCSVファイル・画像ファイル群・縮小画像ファイル群）をダウンロードする。"""
		self.download_csv()
		a_reader = Reader(self.table())
		a_reader.perform()
		os.makedirs(os.path.join(self.attributes().base_directory(), "images"))
		os.makedirs(os.path.join(self.attributes().base_directory(), "thumbnails"))

		self.download_images(self.table().image_filenames())
		self.download_images(self.table().thumbnail_filenames())

		return

File: Java/csv2html/Downloader.java
package csv2html;

import java.awt.image.BufferedImage;
import java.io.File;
import java.util.List;
import utility.ImageUtility;

/**
 * ダウンローダ：CSVファイル・画像ファイル・サムネイル画像ファイルをダウンロードする。
 */
public class Downloader extends IO
{
	/**
	 * ダウンローダのコンストラクタ。
	 * @param aTable テーブル
	 */
	public Downloader(final Table aTable)
	{
		super(aTable);

		return;
	}

	/**
	 * 総理大臣の情報を記したCSVファイルをダウンロードする。
	 */
	public void downloadCSV()
	{
		System.out.println("From: " + super.attributes().csvUrl());
		final List<String> aList = IO.readTextFromURL(super.attributes().csvUrl());
		System.out.println("To: " + super.attributes().baseDirectory());
		super.setCsvFilePath(super.attributes().baseDirectory() + "csv" + File.separator + super.attributes().titleString() + ".csv");
		IO.writeText(aList, super.getCsvFilePath());

		return;
	}

	/**
	 * 総理大臣の画像群をダウンロードする。
	 */
	public void downloadImages()
	{
		final int indexOfImage = this.attributes().indexOfImage();
		this.downloadPictures(indexOfImage);

		return;
	}

	/**
	 * 総理大臣の画像群またはサムネイル画像群をダウンロードする。
	 * @param indexOfPicture 画像のインデックス
	 */
	private void downloadPictures(final int indexOfPicture)
	{
		super.tuples().stream()
		.map(aTuple -> aTuple.values().get(indexOfPicture))
		.forEach(this::downloadPicturesLog);

		return;
	}

	/**
	 * 総理大臣の画像群またはサムネイル画像群をダウンロードする時のログを出力する。
	 * @param aString 各タプルの画像の属性
	 */
	public void downloadPicturesLog(String aString){
		final String anotherString = super.attributes().baseUrl() + aString;
		System.out.println("From: " + anotherString);
		final BufferedImage aBufferedImage = ImageUtility.readImageFromURL(anotherString);
		aString = super.attributes().baseDirectory() + aString;
		ImageUtility.writeImage(aBufferedImage, aString);
		System.out.println("To: " + aString);
	}

	/**
	 * 総理大臣の画像群をダウンロードする。
	 */
	public void downloadThumbnails()
	{
		final int indexOfThumbnail = this.attributes().indexOfThumbnail();
		this.downloadPictures(indexOfThumbnail);

		return;
	}

	/**
	 * 総理大臣の情報を記したCSVファイルをダウンロードして、画像群やサムネイル画像群もダウロードする。
	 */
	public synchronized void run()
	{
		this.makeAssetDir("csv");
		this.makeAssetDir("images");
		this.makeAssetDir("thumbnails");
		this.downloadCSV();
		final Reader aReader = new Reader(super.table());
		aReader.start();
		try {
            aReader.join();
		} catch (final InterruptedException interruptedException) { interruptedException.printStackTrace(); }
		this.downloadImages();
		this.downloadThumbnails();

		return;
	}

	/**
	 * 画像やcsvファイルを格納するディレクトリを生成する。
	 * @param aString 生成するディレクトリ名
	 */
	public void makeAssetDir(String aString){
		aString = super.attributes().baseDirectory() + aString;
		final File aDirectory = new File(aString);
		if (!aDirectory.exists()) { aDirectory.mkdir(); }

		return;
	}
}
